File: src/main/java/com/oycm/algorithm/e/enum_right_log_left/basic/Solution_19.scala
package com.oycm.algorithm.e.enum_right_log_left.basic

object Solution_19 {

  /**
   * 1814. 统计一个数组中好对子的数目 1738
   * https://leetcode.cn/problems/count-nice-pairs-in-an-array/description/
   *
   * rev(x) => rev(123) = 321, rev(120) = 21, 表示将 x 数位反转后得到的结果
   * 好对子定义：
   *  - 0 <= i < j < nums.length
   *  - nums[i] + rev(nums[j]) == nums[j] + rev(nums[i])
   *
   * @param nums 非负整数
   * @return
   */
  def countNicePairs(nums: Array[Int]): Int = {
    /*
    问题等价 nums(i) - rev(nums(i)) = nums(j) - rev(nums(i)) 的对数
    全是 1 的情况下 答案 n * (n - 1) / 2 = 5000000000 < Long.MaxValue

    时间复杂度 O(9n)
    空间复杂度 O(n)
     */
    val mod = 1000000007
    var ans = 0L
    val map = scala.collection.mutable.Map[Int, Int]()
    for (num <- nums) {
      var rev = 0
      var k = num
      while (k > 0) {
        rev = rev * 10 + (k % 10)
        k /= 10
      }
      ans += map.getOrElse(num - rev, 0)
      map(num - rev) = map.getOrElse(num - rev, 0) + 1
    }

    (ans % mod).toInt
  }

  def main(args: Array[String]): Unit = {
    println(countNicePairs(Array(42, 11, 1, 97)))
    println(countNicePairs(Array(13, 10, 35, 24, 76)))
  }

}

File: src/main/java/com/oycm/hot100/matrix/Solution_19.java
package com.oycm.hot100.matrix;

import java.util.ArrayList;
import java.util.List;

public class Solution_19 {

    /**
     * 54. <a href="https://leetcode.cn/problems/spiral-matrix/description/">螺旋矩阵</a>
     *
     * @param matrix
     * @return
     */
    public List<Integer> spiralOrder(int[][] matrix) {
        /*
        给你一个 m 行 n 列的矩阵 matrix ，请按照 顺时针螺旋顺序 ，返回矩阵中的所有元素。
         */
        /*
        标记法
        定义能走的 4 个方向 右下左上，当碰到边界或已经走过的点时，就需要拐弯
         */
        int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int m = matrix.length, n = matrix[0].length;
        int size = m * n;
        List<Integer> ans = new ArrayList<>(size);
        int di = 0;
        int i = 0;
        int j = 0;
        for (int k = 0; k < size; k++) {
            ans.add(matrix[i][j]);
            // 标记为访问
            matrix[i][j] = Integer.MAX_VALUE;
            // 计算下一步位置
            int x = i + dirs[di][0];
            int y = j + dirs[di][1];
            // 看位置是否需要转弯
            if (x < 0 || x >= m || y < 0 || y >= n || matrix[x][y] == Integer.MAX_VALUE) {
                di = (di + 1) % 4;
            }
            i += dirs[di][0];
            j += dirs[di][1];
        }

        return ans;
    }


}
